src/lang-promql/parser/path-finder.ts

```diff
@@ -0,0 +1,94 @@
+// The MIT License (MIT)
+//
+// Copyright (c) 2020 The Prometheus Authors
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+import { NodeType, Subtree } from 'lezer-tree';
+
+function contains(nodeType: NodeType, nodes: (number | string)[]): boolean {
+  let result = false;
+  let i = 0;
+  while (i < nodes.length && !result) {
+    result = nodeType.id === nodes[i] || nodeType.name === nodes[i];
+    i++;
+  }
+  return result;
+}
+
+export function walkBackward(node: Subtree, exit: number) {
+  let result: Subtree | null = node;
+  while (result && result.type.id !== exit) {
+    result = result.parent;
+  }
+  return result;
+}
+
+export function walkThrough(node: Subtree, ...path: (number | string)[]): Subtree | undefined | null {
```

Comment:
That's because the method `iterate` is returning the type `T | undefined`. And during the iteration I can return or a `Subtree` or `null`. So at the end it could return these three values

Reply:
But could you add a comment on what these two falsy values mean semantically for the caller?

Reply:
yeah of course I will add some comments

Reply:
here we go, like that it should explain precisely what is going on for each different value returned